swift/extractor/SwiftOutputRewrite.cpp

```diff
@@ -0,0 +1,318 @@
+// It does not change the original map file as it is dependent upon by the original compiler
+// Returns path to the newly created output file map on success, or None in a case of failure
+static std::optional<std::string> rewriteOutputFileMap(
+    const codeql::SwiftExtractorConfiguration& config,
+    const std::string& outputFileMapPath,
+    const std::vector<std::string>& inputs,
+    std::unordered_map<std::string, std::string>& remapping) {
+  auto newPath = config.tempArtifactDir + '/' + outputFileMapPath;
+
+  // TODO: do not assume absolute path for the second parameter
+  auto outputMapOrError = swift::OutputFileMap::loadFromPath(outputFileMapPath, "");
+  if (!outputMapOrError) {
+    return std::nullopt;
+  }
+  auto oldOutputMap = outputMapOrError.get();
+  swift::OutputFileMap newOutputMap;
+  std::vector<llvm::StringRef> keys;
+  for (auto& key : inputs) {
+    auto oldMap = oldOutputMap.getOutputMapForInput(key);
+    if (!oldMap) {
+      continue;
+    }
+    keys.push_back(key);
+    auto& newMap = newOutputMap.getOrCreateOutputMapForInput(key);
+    newMap.copyFrom(*oldMap);
+    for (auto& entry : newMap) {
+      auto oldPath = entry.getSecond();
+      auto newPath = config.tempArtifactDir + '/' + oldPath;
+      entry.getSecond() = newPath;
+      remapping[oldPath] = newPath;
+    }
+  }
+  std::error_code ec;
+  llvm::SmallString<PATH_MAX> filepath(newPath);
+  llvm::StringRef parent = llvm::sys::path::parent_path(filepath);
+  if (std::error_code ec = llvm::sys::fs::create_directories(parent)) {
+    std::cerr << "Cannot create relocated output map dir: '" << parent.str()
+              << "': " << ec.message() << "\n";
+    return std::nullopt;
+  }
+
+  llvm::raw_fd_ostream fd(newPath, ec, llvm::sys::fs::OF_None);
+  newOutputMap.write(fd, keys);
+  return newPath;
+}
+
+// This is Xcode-specific workaround to produce alias names for an existing .swiftmodule file.
+// In the case of Xcode, it calls the Swift compiler and asks it to produce a Swift module.
+// Once it's done, Xcode moves the .swiftmodule file in another location, and the location is
+// rather arbitrary. Here are examples of such locations:
+// Original file produced by the frontend:
+//  DerivedData/<Project>/Build/Intermediates.noindex/<Project>.build/<BuiltType>-<Target>/<Project>.build/Objects-normal/<Arch>/<ModuleName>.swiftmodule
+// where:
+//  Project: name of a project, target, or scheme
+//  BuildType: Debug, Release, etc.
+//  Target: macOS, iphoneos, appletvsimulator, etc.
+//  Arch: arm64, x86_64, etc.
+//
+// So far we observed that Xcode can move the module into different locations, and it's not
+// entirely clear how to deduce the destination from the context available for the extractor.
+// 1. First case:
+//  DerivedData/<Project>/Build/Products/<BuiltType>-<Target>/<ModuleName>.swiftmodule/<Arch>.swiftmodule
+//  DerivedData/<Project>/Build/Products/<BuiltType>-<Target>/<ModuleName>.swiftmodule/<Triple>.swiftmodule
```

Comment:
How do I read this? The original path on l. 66 can become one of two things in this first case? Either `.../<Arch>.switftmodule` or `.../<Triple>.switftmodule`

Reply:
Originally Xcode asks for `<Arch>/<Name>.swiftmodule` and them moves it into `<Name>.swiftmodule/<Triple>`.
So I guess we can actually ignore the `<Arch>.swiftmodule` case, but it is the format the SPM (Swift Package Manager) uses IIRC.